src/Hintman/Hint/NoNewlineAtFileEnd.hs

```diff
@@ -0,0 +1,59 @@
+module Hintman.Hint.NoNewlineAtFileEnd
+       ( getNoNewlineAtFileEndComments
+       ) where
+
+import Hintman.Core.Hint (Hint, HintType (NoNewlineAtFileEnd), Line (..), simpleSuggestion, toLines)
+import Hintman.Core.PrInfo (ModifiedFile (..))
+import Hintman.Core.Review (Comment (..), createComment)
+
+import qualified Data.Text as T
+
+
+-- | Creates 'Comment's on adding newlines in all modified files.
+getNoNewlineAtFileEndComments :: [ModifiedFile] -> [Comment]
+getNoNewlineAtFileEndComments = mapMaybe getFileNoNewlineAtFileEndComment
+
+-- | Takes a file and returns a comment if file doesn't end with a '\n'.
+getFileNoNewlineAtFileEndComment :: ModifiedFile -> Maybe Comment
+getFileNoNewlineAtFileEndComment mf@ModifiedFile{..} = spawnComment $
+    maybeGetNonEmptyLastLine fileLines diffText
+  where
+    diffText :: Text
+    diffText = decodeUtf8 mfContent
+
+    fileLines :: [Line]
+    fileLines = toLines diffText
+
+    hint :: Hint
+    hint = simpleSuggestion NoNewlineAtFileEnd "No newline at end of file" ""
+
+    -- | Spawns a "NoNewlineAtFileEnd" comment on a given line.
+    spawnComment :: Maybe Line -> Maybe Comment
+    spawnComment = (=<<) (\l -> createComment mf l hint)
+
+-- | Checks if file ends with a newline, and if it does not, returns the last line (to spawn warning on).
+maybeGetNonEmptyLastLine :: [Line] -> Text -> Maybe Line
+maybeGetNonEmptyLastLine mfLines fileContent = do
+    -- 1. Get the last line of the file.
+    lastLine <- viaNonEmpty last mfLines
+
+    -- 2. Return line for warning creation if the last symbol isn't '\n'.
+    maybeLineForWarning lastLine lastSymbol
+  where
+    {- | Checks whether a symbol can be considered a newline.
+    'Nothing' is considered a newline as well, since this means an empty file. -}
+    isNewline :: Maybe Char -> Bool
+    isNewline ls = case ls of
+        Nothing     -> True
+        Just symbol -> symbol == '\n'
+
+    -- | Takes a last symbol of file (if there is any).
+    lastSymbol :: Maybe Char
+    lastSymbol = if T.null fileContent then Nothing
+        else Just $ T.last fileContent
+
+    -- | Returns a line if it's not empty.
+    maybeLineForWarning :: Line -> Maybe Char -> Maybe Line
+    maybeLineForWarning line lastChar
+        | isNewline lastChar = Nothing
+        | otherwise = Just line
```

Comment:
Sorry for the stupid question, but I don't quite get why there is a double check on the newline? Is not it enough to only check the last symbol always? Or you see some cases when it doesn't work and you need to break that into lines and check that too?

Reply:
Talking about the upper-level comment, that's exactly the check for the empty file :) If we don't have any lines in file, we don't want to spawn a warning, even though an empty file (obviously) doesn't have a `\n` symbol at the end.

Thus, firstly we split file contents to lines (which will trim the last `\n` symbol, if any), check that we do have at least one line in file, and only then we can check if `\n` exists at the last line.

Should I expand comments to explain this?

Reply:
To expand the explanations a bit:
- First, we check the last symbol of file (being unaware of `Line`s). Since file can be empty, the return type is `Maybe Char`.
- Then, we split file into lines (since `Line` is an essential concept of spawning a lint). Here an important note is that result of `lines` will trim a `\n` symbol from the last line, meaning that
```haskell
> lines "hello\nworld\n"
["hello","world"]
```
(and that's a reason why we can't just check the last symbol of line, we have to check last symbol of raw content).
- Now, we have an array of lines and the last symbol. We attempt to take the last line, and if there is no lines, the execution quits.
- If we do have the last line, we check that the last symbol is a newline character. At this point, we already know that file is not empty (since we got the line to work with), and thus our `Maybe Char` cannot be `Nothing`. But since we got the character before obtaining this knowledge, we have to process `Nothing` variant as well (I think it's better than raising an error).

As a conclusion, empty file or file ending with `\n` will have no warning, and otherwise the warning will be spawned on the last line of file.

Reply:
Yes, right. But I still think that it's overhead. You can check just on the last character or that the whole content is `""`. What you're describing is very space and time consuming operations, don't really think it worth it :) Let me know if you think that this approach won't work.

Reply:
Well, we still need a `Line` to spawn a comment on, don't we? Or is there a solution which will allow us to avoid work on?

Reply:
Also, as I mentioned in the comment below, we need a last line to create a suggestion: for sample line `end_of_file` the suggestion should be `end_of_file\n`.

Reply:
This is a fair point!
Let me think a bit more on the implementation then :slightly_smiling_face: